src/jtag/drivers/bcm2835gpio.c

```diff
@@ -172,23 +176,46 @@ static void initialize_gpio(enum adapter_gpio_config_index idx)
 
 static bb_value_t bcm2835gpio_read(void)
 {
-	unsigned int shift = adapter_gpio_config[ADAPTER_GPIO_IDX_TDO].gpio_num;
-	uint32_t value = (GPIO_LEV >> shift) & 1;
+	unsigned int shift = adapter_gpio_config[ADAPTER_GPIO_IDX_TDO].gpio_num & 0x1f;
+	uint32_t value = (GPIO_LEV(adapter_gpio_config[ADAPTER_GPIO_IDX_TDO].gpio_num) >> shift) & 1;
 	return value ^ (adapter_gpio_config[ADAPTER_GPIO_IDX_TDO].active_low ? BB_HIGH : BB_LOW);
 
 }
 
 static int bcm2835gpio_write(int tck, int tms, int tdi)
 {
-	uint32_t set = tck << adapter_gpio_config[ADAPTER_GPIO_IDX_TCK].gpio_num |
-			tms << adapter_gpio_config[ADAPTER_GPIO_IDX_TMS].gpio_num |
-			tdi << adapter_gpio_config[ADAPTER_GPIO_IDX_TDI].gpio_num;
-	uint32_t clear = !tck << adapter_gpio_config[ADAPTER_GPIO_IDX_TCK].gpio_num |
-			!tms << adapter_gpio_config[ADAPTER_GPIO_IDX_TMS].gpio_num |
-			!tdi << adapter_gpio_config[ADAPTER_GPIO_IDX_TDI].gpio_num;
-
-	GPIO_SET = set;
-	GPIO_CLR = clear;
+    if ((adapter_gpio_config[ADAPTER_GPIO_IDX_TCK].gpio_num & 32) == 
+        (adapter_gpio_config[ADAPTER_GPIO_IDX_TMS].gpio_num & 32) &&
+        (adapter_gpio_config[ADAPTER_GPIO_IDX_TMS].gpio_num & 32) == 
+        (adapter_gpio_config[ADAPTER_GPIO_IDX_TDI].gpio_num & 32)) {
```

Comment:
Would using ` > 32` here be more "technically correct" than ` & 32` ?

Reply:
yes, that's clearer